## **4.1 Uso de Algoritmos para Aprendizaje Supervisado**
Andrea Mirelle Peralta Ogaz     329516

In [302]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import mean_squared_error
from sklearn.svm import SVR
from sklearn.metrics import r2_score
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsRegressor

### **1. Lectura de archivos utilizando la librería pandas.**

In [303]:
apple_csv = '/Users/andrea.peralta/Curso_Opcion_Tesis/4.1 Algoritmos para Aprendizaje Supervisado/apple_quality.csv'

df = pd.read_csv(apple_csv)

In [304]:
air_csv = '/Users/andrea.peralta/Curso_Opcion_Tesis/4.1 Algoritmos para Aprendizaje Supervisado/AirQuality.csv'

df2 = pd.read_csv(air_csv)

### **2. División de los datos en conjuntos de entrenamiento y prueba.**
Antes de proceder con la división de los datos, se llevó a cabo una visualización de las primeras entradas del conjunto de datos para obtener una previsualización. 

Posteriormente, se verificó el número total de filas y columnas en el conjunto de datos.

In [305]:
df.head(5)

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0     0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1     1 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2     2 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3     3 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4     4  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

    Acidity Quality  
0 -0.491590    good  
1 -0.722809    good  
2  2.621636     bad  
3  0.790723    good  
4  0.501984    good

In [306]:
df2.head(5)

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6         1360       150     11,9           1046   
1  10/03/2004  19.00.00      2         1292       112      9,4            955   
2  10/03/2004  20.00.00    2,2         1402        88      9,0            939   
3  10/03/2004  21.00.00    2,2         1376        80      9,2            948   
4  10/03/2004  22.00.00    1,6         1272        51      6,5            836   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0      166          1056      113          1692         1268  13,6  48,9   
1      103          1174       92          1559          972  13,3  47,7   
2      131          1140      114          1555         1074  11,9  54,0   
3      172          1092      122          1584         1203  11,0  60,0   
4      131          1205      116          1490         1110  11,2  59,6   

       AH  
0  0,7578  
1  0,7255  
2  0,7502  
3  0,7867  
4  0,7888

In [307]:
df = df.drop(columns=['A_id'])

if df.isnull().values.any():
    df = df.dropna()

df = df.drop_duplicates()

In [308]:
df2 = df2.drop(columns=['Date'])
df2 = df2.drop(columns=['Time'])

if df2.isnull().values.any():
    df2 = df2.dropna()

df2 = df2.drop_duplicates()

In [309]:
df["Quality"] = df["Quality"].map({'good': 0, 'bad': 1})  

In [310]:
df2['CO(GT)'] = df2['CO(GT)'].astype(str)  
df2['CO(GT)'] = df2['CO(GT)'].str.replace(',', '.')  

df2['C6H6(GT)'] = df2['C6H6(GT)'].astype(str)  
df2['C6H6(GT)'] = df2['C6H6(GT)'].str.replace(',', '.')  

df2['T'] = df2['T'].astype(str)  
df2['T'] = df2['T'].str.replace(',', '.')  

df2['RH'] = df2['RH'].astype(str)  
df2['RH'] = df2['RH'].str.replace(',', '.')

df2['AH'] = df2['AH'].astype(str)  
df2['AH'] = df2['AH'].str.replace(',', '.')  

In [311]:
df.head(5)

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness   Acidity  \
0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840 -0.491590   
1 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530 -0.722809   
2 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033  2.621636   
3 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761  0.790723   
4  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849  0.501984   

   Quality  
0        0  
1        0  
2        1  
3        0  
4        0

In [312]:
df2.head(5)

CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
0    2.6         1360       150     11.9           1046      166   
1      2         1292       112      9.4            955      103   
2    2.2         1402        88      9.0            939      131   
3    2.2         1376        80      9.2            948      172   
4    1.6         1272        51      6.5            836      131   

   PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH      AH  
0          1056      113          1692         1268  13.6  48.9  0.7578  
1          1174       92          1559          972  13.3  47.7  0.7255  
2          1140      114          1555         1074  11.9  54.0  0.7502  
3          1092      122          1584         1203  11.0  60.0  0.7867  
4          1205      116          1490         1110  11.2  59.6  0.7888

In [313]:
print(df.shape[0], "," , df.shape[1])

4000 , 8


In [314]:
print(df2.shape[0], "," , df2.shape[1])

9326 , 13


Se utilizó df.iloc en Python para seleccionar y acceder a elementos específicos en un DataFrame:
1. En **X = df.iloc[:, :-1]**: X se crea seleccionando todas las filas (:) y las primeras siete columnas (índices de columna 0 a 7) de df.

2. Para **y = df.iloc[:, -1]**: y se crea seleccionando todas las filas (:) y solo la octava columna (índice de columna 7) de df.

In [315]:
# Obtener todas las columnas excepto la última
X = df.iloc[:, :-1]

# Obtener solo la última columna
y = df.iloc[:, -1]

In [316]:
A = df2.iloc[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12]]

b = df2.iloc[:, 10]

Nuevamente se verificó el número total de filas y columnas en el conjunto de datos, para visualizar el cambio.

In [317]:
X.shape

(4000, 7)

In [318]:
y.shape

(4000,)

Se utilizó la función **train_test_split** de la biblioteca **scikit-learn** para llevar a cabo la división de un conjunto de datos en conjuntos de entrenamiento y prueba. En esta implementación, se asignó el 80% de los datos para el conjunto de entrenamiento y el 20% para el conjunto de prueba.

In [319]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42, stratify=y)

In [320]:
A_train, A_test, b_train, b_test = train_test_split(A, b, test_size=0.20)

In [321]:
X_train.shape

(3200, 7)

In [322]:
A_train.shape

(7460, 12)

In [323]:
X_test.shape

(800, 7)

In [324]:
A_test.shape

(1866, 12)

In [325]:
pca_obj = PCA(n_components=11)
pca_obj.fit(A_train)
pca_obj.explained_variance_

array([5.54882786e+05, 1.53259973e+05, 5.67904301e+04, 2.96911531e+04,
       1.87569373e+04, 7.82909318e+03, 5.16764262e+03, 4.37565121e+03,
       2.42905262e+03, 4.78233759e+02, 6.49603390e+01])

In [326]:
#Tansformación de datos
#Generar conjuntos de entrenamiento y prueba
new_A_train = pca_obj.transform(A_train)
new_A_test = pca_obj.transform(A_test)

new_A_train.shape

(7460, 11)

### **3. Escalado de las Variables Predictoras al Intervalo 0,1**

In [327]:
scaler = MinMaxScaler(feature_range=(0, 1))

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [328]:
scaler = MinMaxScaler(feature_range=(0, 1))

A_train_scaled = scaler.fit_transform(A_train)

A_test_scaled = scaler.transform(A_test)

### **4. Aplicación del Algoritmo SVM y el KNN para Clasificar el Conjunto de Datos "Apple Quality"**

In [329]:
svm_model = SVC()
svm_model.fit(X_train_scaled, y_train)

y_pred_svm = svm_model.predict(X_test_scaled)

svm_accuracy = accuracy_score(y_test, y_pred_svm)

print("Precisión del modelo SVM:", svm_accuracy)

svm_conf_matrix = confusion_matrix(y_test, y_pred_svm)
print("Matriz de confusión del modelo SVM:")
print(svm_conf_matrix)

Precisión del modelo SVM: 0.895
Matriz de confusión del modelo SVM:
[[360  41]
 [ 43 356]]


In [330]:
knn_model = KNeighborsClassifier()
knn_model.fit(X_train_scaled, y_train)

y_pred_knn = knn_model.predict(X_test_scaled)

knn_accuracy = accuracy_score(y_test, y_pred_knn)

print("Precisión del modelo KNN:", knn_accuracy)

knn_conf_matrix = confusion_matrix(y_test, y_pred_knn)
print("Matriz de confusión del modelo KNN:")
print(knn_conf_matrix)

Precisión del modelo KNN: 0.88
Matriz de confusión del modelo KNN:
[[358  43]
 [ 53 346]]


#### **Conclusión SVM**
El modelo SVM tiene una precisión del 89.5%. Esta es una métrica común para evaluar modelos de clasificación y se calcula como la proporción de predicciones correctas sobre el total de predicciones realizadas. Por lo tanto, en este caso, el 89.5% de las predicciones del modelo SVM fueron correctas.
-    Verdaderos positivos (TP): 360 observaciones fueron clasificadas correctamente como "buenas".
-    Falsos positivos (FP): 41 observaciones fueron clasificadas incorrectamente como "buenas" cuando en realidad eran "malas".
-    Verdaderos negativos (TN): 356 observaciones fueron clasificadas correctamente como "malas".
-    Falsos negativos (FN): 43 observaciones fueron clasificadas incorrectamente como "malas" cuando en realidad eran "buenas".

#### **Conclusión KNN**
Precisión del modelo: El modelo KNN tiene una precisión del 88%. En términos generales, una precisión del 88% se considera bastante buena.
-    Verdaderos positivos (TP): Hay 358 observaciones que fueron clasificadas correctamente como positivas (en este caso, "buenas").
-    Falsos positivos (FP): Se clasificaron incorrectamente 43 observaciones como positivas cuando en realidad eran negativas (en este caso, "malas").
-    Verdaderos negativos (TN): 346 observaciones se clasificaron correctamente como negativas.
-    Falsos negativos (FN): Se clasificaron incorrectamente 53 observaciones como negativas cuando en realidad eran positivas.

### **5. Aplicación del Algoritmo SVM y el KNN para Clasificar el Conjunto de Datos "Air Quality"**

In [331]:
# Ajuste del modelo SVM para regresión
svm_regressor = SVR(kernel='rbf')
svm_regressor.fit(A_train_scaled, b_train)

# Predicciones con SVM
y_pred_svm = svm_regressor.predict(A_test_scaled)

# Evaluación del modelo SVM
mse_svm = mean_squared_error(b_test, y_pred_svm)
r2_svm = r2_score(b_test, y_pred_svm)

print("Error cuadrático medio (SVM):", mse_svm)
print("Coeficiente de determinación (R^2) (SVM):", r2_svm)

Error cuadrático medio (SVM): 23.74658595205142
Coeficiente de determinación (R^2) (SVM): 0.9883260383394016


In [338]:
k = int(np.sqrt(new_A_train.shape[0]))
knn_regressor = KNeighborsRegressor(n_neighbors=5)
knn_regressor.fit(A_train_scaled, b_train)

A_test_scaled = A_test_scaled.astype(float)

# Predicciones con KNN
b_pred_knn = knn_regressor.predict(A_test_scaled)

# Evaluación del modelo KNN
mse_knn = mean_squared_error(b_test, b_pred_knn)
r2_knn = r2_score(b_test, b_pred_knn)

print("Error cuadrático medio (KNN):", mse_knn)
print("Coeficiente de determinación (R^2) (KNN):", r2_knn)

Error cuadrático medio (KNN): 5.604095176848874
Coeficiente de determinación (R^2) (KNN): 0.9972449937700949


#### **Conclusión SVR**
Un MSE de 23.75 es bastante bajo, lo que indica que el modelo SVM tiene un buen rendimiento y tiende a hacer predicciones precisas en promedio.
El coeficiente de determinación (R^2) mide la proporción de la varianza en la variable dependiente que se puede predecir a partir de la variable independiente(s). En este caso, un valor de 0.9883 indica que el modelo SVM explica el 98.83% de la varianza en la variable objetivo, lo cual es muy alto y sugiere un buen ajuste del modelo.

#### **Conclusión KNN**
Un MSE de 5.604 es bastante bajo, lo que indica que el modelo KNN tiene un rendimiento excelente y tiende a hacer predicciones muy precisas en promedio.
Un valor de 0.997 para el coeficiente de determinación indica que el modelo KNN explica el 99.7% de la varianza en la variable objetivo, lo cual es excepcionalmente alto y sugiere un ajuste del modelo casi perfecto.